docs/proposals/20190610-machine-states-preboot-bootstrapping.md

```diff
@@ -0,0 +1,446 @@
+      * [Certificates and tokens are exposed in plaintext](#certificates-and-tokens-are-exposed-in-plaintext)
+      * [Bootstrap data cannot be merged](#bootstrap-data-cannot-be-merged)
+      * [MachineClass is deprecated and will be revisited later](#machineclass-is-deprecated-and-will-be-revisited-later)
+  * [Design Details](#design-details)
+    * [Test Plan](#test-plan)
+    * [Graduation Criteria](#graduation-criteria)
+    * [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+    * [Version Skew Strategy](#version-skew-strategy)
+  * [Implementation History](#implementation-history)
+  * [Drawbacks](#drawbacks)
+  * [Alternatives](#alternatives)
+      * [Object References, Templates, MachineSets and MachineDeployments](#object-references-templates-machinesets-and-machinedeployments-1)
+
+## Glossary
+
+- **Cluster API**: Unless otherwise specified, this refers to the project as a whole.
+- **Machine**: the Custom Resource for Kubernetes
+- **Server/Instance/Host**: the infrastructure that backs a Machine
+- **Bootstrapping**: turning a server into a Kubernetes node
+
+## Summary
+
+This proposal outlines splitting up the Machine data model into two new providers/controllers along with the generic Machine controller. The two new providers are the Infrastructure provider and the Bootstrap provider. The infrastructure provider handles provisioning infrastructure (cloud, bare metal, virtual) and the bootstrap provider handles Kubernetes bootstrapping. This change will improve separation of concerns for the Machine controller and gives more degrees of freedom to the user.
+
+In Cluster API v1alpha1, users can create Machine custom resources. Cluster API components react to a new Machine by provisioning infrastructure (typically a physical or virtual machine) and then bootstrapping Kubernetes on the provisioned server, making it a Kubernetes Node. Cluster API’s current architecture mandates that a single “provider” handle both the infrastructure provisioning and the Kubernetes bootstrapping.
+
+## Motivation
+- A singular provider that combines infrastructure provisioning and Kubernetes bootstrapping is too restrictive. Many current v1alpha1 providers target single cloud providers. If a user wants to install a specific Kubernetes distribution (Kubernetes, OpenShift, Rancher, etc.), it is left to the provider to add support for it. In practical terms, we are not able to guarantee that a user can install the same distribution regardless of infrastructure environment.
+- Both users and developers are confused by the code organization split where all the controllers live in the cluster-api repository, but some controllers are part of the generic cluster-api pod (MachineDeployment, MachineSet), while other controllers are part of each provider’s pod (Cluster, Machine). We would like for the code location to correspond to the pod, which hopefully will reduce both user and developer confusion.
```

Comment:
s/Some generic controllers like MachineDeployment and Machineset live in the cluster-api repo, while some provider specific controllers like Cluster and Machine , live in the provider specific repo. 

Deployment of controllers depends on choice. For example, in OpenShift, all controllers are deployed in single pod.

Reply:
I'm not sure if this actually has to change. The motivation is meant to define the current state of the world for `kubernetes-sigs` repositories and code organization. While we're aware that other implementation exist outside, the main focus is to provide a holistic view limited the subproject's repositories.

Reply:
I don't see direct implication of the second motivation point wrt. `Machine States & Preboot Bootstrapping` title. Though I understand it's important to provide layout and manifests that reflect structure of the new to-be-done cluster-api plane and relation between individual components. And also that's up to each consumer to decide if all the components of the cluster-api plane will live inside the same pod (or, will be distributed among multiple pods/deployments). Given the fact, implication of the second motivation item transcends beyond cluster-api, making what the right view might look like on cluster-api repository not easily predictable and questionable. Still, it's important to provide some default view.

As I read the second item, it does not sufficiently express its intention in the scope of the proposal. Does the first part reflect the following rephrasing?

```
Both users and developers are confused by the code organization split where generic controllers (MachineDeployment, MachineSet) live fully inside cluster-api repository while other (Cluster, Machine) are partially implemented in cluster-api repository but vendored in provider repositories that implement provider specific bits. Thus, giving existence to separate pods which both deploy all controllers instead of a single one.
```

If so, can we be more specific in what `We would like for the code location to correspond to the pod, which hopefully will reduce both user and developer confusion.` actually means? Is it to create completely new layout where machine controller will become fully general without injecting actuator implementation? And, actuators will be turned into separate controllers with their own life-cycle where a relation between machine life-cycle and actuator will be provided? Or, machine controllers will still be fully vendored in provider repositories as it is today and the intention is to just change currently provided manifests in the cluster-api repository?

Reply:
@ingvagabund @vikaschoudhary16 I think this point is just poorly phrased and I'll need to rework it a little later or tomorrow. Thanks for the great feedback. I read it again multiple times and I think it's unclear what we initially wanted to communicate.

How does that sound?

Reply:
I understand the intention of this to be that a clearer separation of concerns between infrastructure and bootstrap providers will be reflected in code organisation and thus easier for developers to grok. It's the `pod` bit here that confuses me: that is a deployment issue, rather than a code organisation issue.

Reply:
To be honest I thought the wording made sense, could be refined.  However I want to make certain we are not conflating ideas.  

> Deployment of controllers depends on choice. For example, in OpenShift, all controllers are deployed in single pod. Vendoring of cluster-api is cumbersome though. 

Is orthogonal.

Reply:
Reworded, ptal